program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,390 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+type SignedTrackerRunningSum = i64;
+type UnsignedTrackerRunningSum = u64;
 
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Status {
+    Invalid = 0,
+    Valid   = 1,
+    Pending = 2,
+}
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+mod track_helpers {
+    use crate::error::OracleError;
+    use crate::utils::try_convert;
+
+    ///gets the index of the next entry
+    pub fn get_next_entry(current_entry: usize, num_entries: usize) -> usize {
+        (current_entry + 1) % num_entries
+    }
+    ///gets the index of the previou entry
+    pub fn get_prev_entry(current_entry: usize, num_entries: usize) -> usize {
+        (current_entry - 1) % num_entries
+    }
+    ///Given a time, return the entry of that time
+    pub fn time_to_entry(
+        time: i64,
+        num_entries: usize,
+        granuality: i64,
+    ) -> Result<usize, OracleError> {
+        Ok(try_convert::<i64, usize>(time / granuality)? % num_entries)
+    }
+
+    ///returns the remining time before the next multiple of GRANULARITY
+    pub fn get_time_to_entry_end(time: i64, granuality: i64) -> i64 {
+        (granuality - (time % granuality)) % granuality
```

Comment:
I think this was already discussed but why is`(granuality - (time % granuality))` instead of `(granuality - (time % granuality)) % granuality` not valid?

Reply:
it won't produce the correct output when time is divisible by 0